### DISCUSSION TITLE
fitting method for high-order force constants

### DISCUSSION THREAD

**Discussion by rfhari** [ttadano/alamode#233] [Q&A] [answered]
Hi @ttadano, I'm interested in calculating second- and third-order force constants. From the [documentation](https://alamode.readthedocs.io/en/latest/almdir/formalism_alm.html), I understand that ALAMODE uses the ordinary least squares method (in the ALM module) to calculate these force constants. However, I was wondering if ALAMODE fits all the interatomic force constants (IFCs) simultaneously, or if it fits them iteratively based on the residuals, as shown in Equations 26–28 of the [TDEP documentation](https://tdep-developers.github.io/tdep/program/extract_forceconstants/). I couldn't find an explicit mention of this in the ALAMODE documentation. Could you please clarify this? 

Thanks! 

PS: Based on the sensing matrix of Eq 3 of this [paper](https://journals.aps.org/prl/pdf/10.1103/PhysRevLett.113.185501) on CSLD, I think, unlike TDEP, Alamode fits is it simultaneously. But I'm not really sure.

**Comment by ttadano** [comment, marked as answer]
Both options are available. If you do not specify any options, all force constants will be fitted simultaneously. If you set the `FC2XML` option in the `&optimize` field, the second-order IFCs are fixed to the values in the given file, and only the anharmonic IFCs are fitted.

**Reply by ejmeitz** [reply]
When using the FC2XML flag I get an error. I've attached the input files and the log file from the run. Based on where the log-file ends I think the program is trying to read the xml file I have. I created the XML file using the ALM python library. A couple places `stoi` could be called but really hard to tell. Happy to debug this and open an issue, but I'm not sure I have the tools.

```
terminate called after throwing an instance of 'std::invalid_argument'
  what():  stoi
```

EDIT: Building the `alm` executable inside of ALM instead of ALAMODE I now get a more useful error: 
```
Error in get_value_from_xml
The following entry could not be found in the XML file : Data.ForceConstants.HarmonicUnique.NFC2
```
Pretty sure this happens because I am manually loading in my own IFC's instead of calculating them with ALM first. So I dont end up with the unique IFCs I just have entries for all the possible IFCs for the primitive cell. Is there any way to generate the FCXML file in a way that will match the expected format of alm? Lookin at the source code [here](https://github.com/ttadano/ALM/blob/develop/python/alm/fcsxml.py#L11) the XML key (HarmonicUnique) is not written by the fcsxml.py but is expected by the xml parser in ALM/ALAMODE ([here](https://github.com/ttadano/ALM/blob/f1d668fdee66e7e7218a04c88daf19d0e14fce0c/src/constraint.cpp#L2434))?

[lj_alm.zip](https://github.com/user-attachments/files/17708512/lj_alm.zip)
